runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
-      iree_wait_one(&timepoint->timepoint.host_wait, deadline_ns);
-  if (!iree_status_is_ok(status)) {
-    iree_hal_semaphore_cancel_timepoint(&semaphore->base, &timepoint->base);
-  }
-  iree_hal_hip_timepoint_pool_release(semaphore->timepoint_pool, 1, &timepoint);
-  if (!iree_status_is_ok(status)) {
-    IREE_TRACE_ZONE_END(z0);
-    return status;
+    return iree_ok_status();
   }
+  iree_status_t status = iree_status_clone(semaphore->failure_status);
+  if (iree_status_is_ok(status)) {
+    iree_hal_hip_util_tree_node_t* node =
+        iree_hal_hip_util_tree_get(&semaphore->event_queue.tree, value);
+
+    if (node == NULL) {
+      status = iree_hal_hip_util_tree_insert(&semaphore->event_queue.tree,
+                                             value, &node);
+      if (iree_status_is_ok(status)) {
+        iree_hal_hip_semaphore_queue_item_t* item =
+            (iree_hal_hip_semaphore_queue_item_t*)
+                iree_hal_hip_util_tree_node_get_value(node);
+        item->cpu_event = NULL;
+        item->work_item = NULL;
+      }
+    }
 
-  iree_slim_mutex_lock(&semaphore->mutex);
-  if (semaphore->current_value >= IREE_HAL_SEMAPHORE_FAILURE_VALUE) {
-    status = iree_make_status(IREE_STATUS_ABORTED);
+    if (iree_status_is_ok(status)) {
+      iree_hal_hip_event_t* event =
+          ((iree_hal_hip_semaphore_queue_item_t*)
+               iree_hal_hip_util_tree_node_get_value(node))
+              ->event;
+      if (!event) {
+        status = iree_hal_hip_event_pool_acquire(
+            event_pool, 1,
+            &((iree_hal_hip_semaphore_queue_item_t*)
+                  iree_hal_hip_util_tree_node_get_value(node))
+                 ->event);
+        if (iree_status_is_ok(status)) {
+          event = ((iree_hal_hip_semaphore_queue_item_t*)
+                       iree_hal_hip_util_tree_node_get_value(node))
+                      ->event;
+          status = IREE_HIP_CALL_TO_STATUS(
+              semaphore->symbols,
+              hipEventRecord(iree_hal_hip_event_handle(event),
+                             dispatch_stream));
+        }
+      }
+      if (event) {
+        iree_hal_hip_event_retain(event);
+      }
+      *out_hip_event = event;
+    }
   }
   iree_slim_mutex_unlock(&semaphore->mutex);
 
   IREE_TRACE_ZONE_END(z0);
+
   return status;
 }
 
-iree_status_t iree_hal_hip_semaphore_multi_wait(
-    const iree_hal_semaphore_list_t semaphore_list,
-    iree_hal_wait_mode_t wait_mode, iree_timeout_t timeout,
-    iree_arena_block_pool_t* block_pool) {
-  if (semaphore_list.count == 0) return iree_ok_status();
-
-  if (semaphore_list.count == 1) {
-    // Fast-path for a single semaphore.
-    return iree_hal_semaphore_wait(semaphore_list.semaphores[0],
-                                   semaphore_list.payload_values[0], timeout);
-  }
-
+static iree_status_t iree_hal_hip_semaphore_query_locked(
+    iree_hal_hip_semaphore_t* semaphore, uint64_t* out_value) {
   IREE_TRACE_ZONE_BEGIN(z0);
```

Comment:
queries likely shouldn't be traced (again worth checking the traces - add as much detail as required and no more, as traces are a commons and if everyone adds everything then they become useless - have to be judicious on anything that is not initialization/deinitialization-time stuff and that may be called thousands/millions of times)